[MRG + 2] ENH RobustScaler

untom opened the pull request:
This PR adds `RobustScaler` and `robust_scale` as alternative to `StandardScaler` and `scale`. They use robust estimates of data center/scale (median & interquartile range), which will work better for data with outliers.

Most of this was discussed before in #2514 (and even older commits). I separated this out so it can be merged as-is.

I originally wanted to submit this only after #3639 was merged, but sending the PR now allows it to be discussed concurrently (if either this or #3639 get merged, I'll of course update the other commit).

untom committed:
ENH Added RobustScaler

ragv commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,259 @@ def inverse_transform(self, X, copy=None):
+        ----------
+        X : array-like or CSR matrix.
+            The data used to scale along the specified axis.
+        """
```
Could you add a line `check_is_fitted(self, 'scale_')`. This is to raise `NotFittedError` when this method is called before `fit`...

untom replied:
done

ragv commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,259 @@ def inverse_transform(self, X, copy=None):
+        if self.with_centering:
+            check_is_fitted(self, 'center_')
+        if self.with_scaling:
+            check_is_fitted(self, 'scale_')
```
Since [you are setting](https://github.com/scikit-learn/scikit-learn/pull/4125/files#diff-5ebddebc20987b6125fffc893f5abc4cR515) both these parameters, could you instead use `check_is_fitted(self, ['scale_', 'center_'], all_or_any=any)` :)

untom replied:
I decided on not setting the parameters instead, as IIRC in a similar PR someone pointed out that they don't like that.

jnothman commented:
Test failure:

AttributeError: 'RobustScaler' object has no attribute 'center_'

untom committed:
Check if parameters are fitted

untom commented:
Fixed now, sorry that it even happened, I only ran the `sklearn.preprocessing` specific tests before.

agramfort commented:
is this ready to review?

untom commented:
Yes it is

untom commented:
Pinging @amueller  @ogrisel  Maybe you want to have a look?

amueller commented:
Yes, but probably only after he beta release is cut on Friday.

untom commented:
ping @amueller is now a good time to remind you about this PR?

amueller commented:
It is always a good time to remind me ;) I'll try to have a look.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+
+    Parameters
+    ----------
+    interquartile_scale: float or string in  ["normal" (default), ],
```
please add a space before " :"

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+    interquartile_scale: float or string in  ["normal" (default), ],
+           The interquartile range is divided by this factor. If
+           `interquartile_scale` is "normal", the data is scaled so it
+           approximately reaches unit variance. This converge assumes Gaussian
```
Sorry, I have a hard time parsing that sentence. what is meant by "converge" here?

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+    def _check_array(self, X, copy):
+        """Makes sure centering is not enabled for sparse matrices."""
+        X = check_array(X, accept_sparse=('csr', 'csc'),
+                        copy=copy, ensure_2d=False)
```
We remove ``warn_if_not_float`` and you can just do ``dtype=[np.float64]`` in ``check_array`` now.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+            used for later scaling along the features axis.
+        """
+        if sparse.issparse(X):
+            raise TypeError("RobustScaler cannot be fitted on sparse inputs")
```
I'm not sure I understand. so we can not fit on sparse data, but we can transform it? And if we can not fit, why does the docstring say "array-like or CSR"?

untom replied:
This is actually something I haven't thought out very well. Fitting a RobustScaler on CSR/CSC might be tricky/expensive, so I didn't implement it (the docstring you pointed out will be corrected to reflect this).

However in theory, one could transform a sparse matrix, but that runs the risk of destroying the sparsity pattern. So the options are:
 (1) only transform CSR/CSC when we are only scaling, but not centering (throw an exception if a CSR matrix is passed and `self.with_centering` is True) => This is what `StandardScaler` does.
 (2)  Only throw an exception if the new center isn't 0 
 (3) Don't accept CSC/CSR at all.

I'm leaning towards option (2) since it's the most general, but (1) is the most consistent -- and (3) is the least complicated. Thoughts?

amueller replied:
What does it mean to compute the percentiles and scale if we don't subtract the mean? Does that assume the mean is zero? That is more a question of what ``with_centering`` should do I'm not entirely sure it was a great API for ``StandardScaler`` but maybe you thought more about this than I did.
2) seems to make the most sense.
If we want to do 2), do we actually have a ``with_centering`` parameter at all?

I'd be fine with bailing and doing 3. In this case, do we want / need ``with_centering``?

untom replied:
Think you're right. Using `with_centering=False` only makes sense for sparse data. But then again, a RobustScaler can't be fitted on sparse data, so it probably doesn't even make sense to allow it to be transformed that way. It's a very unusual usecase. It's probably best to remove `with_centering` for robust scalers.

amueller replied:
Ok. I am always for simpler interfaces and this seems indeed unusual. I'd be happy with adding this just for dense data for the moment but 2) would be good, too.

untom replied:
I agree that simpler interfaces are nice. the "completest" approach would be:  remove `with_centering`, forbid fitting sparse matrices, but still allow sparse matrices to be transformed if center == 0

amueller replied:
well you could also allow fitting with sparse data if center == 0, right?

untom replied:
I don't know how to calculate the median nor the IQR of sparse data, so I cannot fit sparse data.

amueller replied:
never mind then ^^

amueller replied:
Well it is possible to correct for the zeros, right? But yeah it is more tricky so let's not do that now.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+            scale[-np.isfinite(scale)] = 1.0
+        return scale
+
+    def fit(self, X, y=None, copy=None):
```
I don't think ``fit`` should have a ``copy`` parameter, I think we are trying to remove these, as there is already an ``__init__`` parameter.

untom replied:
Does the same go for e.g. `transform` and  `inverse_transform`?

amueller replied:
I would think so. I see that we haven't deprecated them in ``StandardScaler``, so I don't have a strong opinion. I'm not sure how useful they are. To me they mostly seem to clutter the API. If you like you can leave them.

untom replied:
I just tried staying close to the `StandardScaler` API, but I don't have a preference either way.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+    Parameters
+    ----------
+    interquartile_scale: float or string in  ["normal" (default), ],
+           The interquartile range is divided by this factor. If
```
I am not sure I understand the parametrization here. So we are always using 1/4 3/4 quantiles and then scale them? Is that a standard practice? [sorry I rarely do robust statistics]

untom replied:
Not sure if you're confused about the `interquartile_scale` parameter or picking the IQR as the range to scale in. In short, IQR is the most common range, yes.

To clear up the confusion about the `interquartile_scale`: it is used to make our estimate of the IQR into an (unbiased) estimate of the variance of the data.  For example, if the underlying data is a Normal distribution, we know that IQR = 1.349 * sigma^2.  The 1.349 is the `interquartile_scale`. So by dividing the IQR by that value and afterwards scaling the data by the resulting factor, we'll get data that has a variance ~ 1.  We could calculate the correct scale to use for any quantile range, not just the IQR.  However, for this to work we need to make assumptions about the underlying data distribution. 

I just tested it, and actually for this to work, the data must follow the assumed distribution pretty closely. To be honest, I think the `interquartile_range` should be removed, it just adds needless complexity.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+        return scale
+
+    def fit(self, X, y=None, copy=None):
+        """Compute the mean and std to be used for later scaling.
```
Median and quantiles?

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+        return self
+
+    def transform(self, X, y=None, copy=None):
+        """Perform standardization by centering and scaling
```
Is "centering" a good verb here? I have no strong opinion, though.

untom replied:
Now that you mention it, I actually  think "standardization" is actually the bad verb in this case (since it's meaning is usually mean-centering and std-scaling) :-D

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +397,261 @@ def inverse_transform(self, X, copy=None):
+
+    Parameters
+    ----------
+    X : array-like or CSR matrix.
```
Doesn't work with CSR, right?

amueller commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -455,6 +457,96 @@ def test_scale_function_without_centering():
+    X_scaled = scaler.fit(X).transform(X, copy=True)
+
+    assert_array_almost_equal(np.median(X_scaled, axis=0), 5 * [0.0])
+    assert_array_almost_equal(X_scaled.std(axis=0)[0], 0)
```
For normal data, the std of the other axis should be 1, right? I guess that might be not enough data for that?

amueller commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -455,6 +457,96 @@ def test_scale_function_without_centering():
+
+
+def test_robust_scaler_iqr_scale():
+    """Does iqr scaling achieve approximately std= 1 on Gaussian data?"""
```
ah, ok, there is a separate test for that ^^ nice

amueller commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -455,6 +457,96 @@ def test_scale_function_without_centering():
+
+
+def test_robust_scaler_zero_variance_features():
+    """Check min max scaler on toy data with zero variance features"""
```
RobustScaler

amueller commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -455,6 +457,96 @@ def test_scale_function_without_centering():
+
+def test_robust_scaler_iqr_scale():
+    """Does iqr scaling achieve approximately std= 1 on Gaussian data?"""
+    rng = np.random.RandomState(42)
```
Maybe a slightly better test would be to add some outliers to the data, fit the scaler, then remove the outliers, transform, and see that std is 1?

amueller commented:
If you like you could add a plot_ example that shows samples from a 2d gaussian with outliers, how it gets transformed using StandardScaler (into something weird) and how using RobustScaler (a ball + outliers). Not required but I think this would be very illustrative.

Apart from my minor nitpicks this looks great. Sorry for the lack of feedback previously.

untom commented:
> I am not sure I understand the parametrization here. So we are always using 1/4 3/4 quantiles and then scale them? Is that a standard practice? [sorry I rarely do robust statistics]

Long answer outside of the affected code lines so the answer stays visible for future reviewers: Yes, using the 0.25-0.75  (the "interquartile range") is the most common option, although others might make sense in some occasions (e.g. [Wikipedia](http://en.wikipedia.org/wiki/Robust_measures_of_scale) also meantions the interdecale range 0.1-0.9 as relevant). We could make the quantiles a user-definable parameter that just defaults to 0.25-0.75, to cover the more general case. The only question is if it's worth the added complexity for the user. I've personally never needed anything more than IQR, but maybe having the option might be nice?

untom committed:
Fix documentation errors.

untom committed:
Remove warn_if_not_float

untom committed:
Remove 'copy' parameter form RobustScaler functions

untom committed:
Fix documentation in test case

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +396,256 @@ def inverse_transform(self, X, copy=None):
+                self.scale_[-np.isfinite(self.scale_)] = 1.0
+        return self
+
+    def transform(self, X, y=None, copy=None):
```
I don't have a strong preference, in general I like to err on the side of simpler interfaces.

amueller commented:
As I just said wrt the copy parameter, I'd like to err on the side of simpler interfaces, so less parameter unless necessary. we can always add more later. My confusion was mostly cause by me reading "quartile" as "quantile". I have never heard the word quartile or the shorthand IQR before, and it would be nice to give a very short explanation and maybe a wikipedia reference.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +396,256 @@ def inverse_transform(self, X, copy=None):
 
 
+class RobustScaler(BaseEstimator, TransformerMixin):
+    """Standardize features by removing the median and scaling to IQR.
```
You use the word "standardize" here and also in the docstring of the function. Maybe not good as discussed below. The other scaler is called "StandardScaler" so maybe just say "Scale features"? I don't know, and that is anyhow maybe to much nitpicking ;)

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +396,256 @@ def inverse_transform(self, X, copy=None):
+    """Standardize features by removing the median and scaling to IQR.
+
+    Centering and scaling happen independently on each feature (or each
+    sample, depending on the `axis` argument) by computing the relevant
```
Maybe after this sentence one sentence "Interquantile range (IQR) is the range between 1st and 3rd quantile, that is between the 25th and 75th percentile" or something similar?

untom committed:
Export robust_scale and RobustScaler

untom committed:
Improved documentation of robust scaler

untom committed:
Fixed decprecation warning

untom committed:
Changed outdated test

untom committed:
Fixed removal of copy-parameter

untom committed:
Removed superfluous testcase for robust scaler

untom committed:
Remove interquartile_scale parameter

untom committed:
Added robust scaling example

untom commented:
I decided to remove the `interquartile_scale` parameter, after making some tests and seeing that it really doesn't work well when the underlying distribution isn't gaussian.

I also followed your advice and added a `plot_robust_scaling.py` example file.

untom committed:
Fix robust_scaling test

amueller commented:
travis failure unrelated, I restarted it

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +399,239 @@ def inverse_transform(self, X, copy=None):
+        return scale
+
+    def fit(self, X, y=None):
+        """Compute the mean and quantiles to be used for scaling.
```
Median?

amueller commented:
you didn't push the example, did you?

untom committed:
Fixed documentation oversight.

untom committed:
robust_scaling example

untom commented:
Whoops, I thought I did -- sorry, it's pushed now.

untom commented:
pinging @amueller : is any additional input from my side needed?

amueller commented on examples/preprocessing/plot_robust_scaling.py:
```diff
@@ -0,0 +1,80 @@
+
+
+# Plot data (note we zoom in to the data center, so the outlier
+# data point can't be seen in this plot).
```
Can we not do it in a way that we can see the points? For a person looking at the docstring and the plot, this is not obvious.

untom replied:
That's a bit tricky, because that would require zooming out so far that you don't see a difference between robust and normal scaling. Instead, I now plot the original data on a separate subplot, as well.

amueller commented on examples/preprocessing/plot_robust_scaling.py:
```diff
@@ -0,0 +1,80 @@
+
+
+# Scale data
+s = StandardScaler()
```
we try to avoid single-letter names and shorthands. Maybe ``standard`` and ``robust``?

amueller commented:
Apart from my nitpicks on the example, this looks good to me. Maybe @ogrisel has time to review?

amueller commented:
Or @agramfort ?

untom committed:
Improved example

amueller commented:
Sounds good

agramfort commented on doc/modules/preprocessing.rst:
```diff
@@ -145,6 +145,16 @@ full formula is::
+Scaling data with outliers
+--------------------------
+If your data contains many outliers, scaling using the mean and variance
+of the data does sometimes not work very well. In these cases, you can use
```
is likely to not work very well

agramfort commented on doc/modules/preprocessing.rst:
```diff
@@ -145,6 +145,16 @@ full formula is::
+of the data does sometimes not work very well. In these cases, you can use
+:func:`robust_scale` and :class:`RobustScaler` as drop-in replacements
+instead, which use more robust estimates for the center and range of your
+data.
```
as drop-in replacements instead, which use more robust estimates for the center and range of your data. ...

->

. It uses robust estimates for the center and range of your data.

agramfort commented on doc/modules/preprocessing.rst:
```diff
@@ -145,6 +145,16 @@ full formula is::
 
+
+Scaling data with outliers
+--------------------------
```
insert empty line

agramfort commented on examples/preprocessing/plot_robust_scaling.py:
```diff
@@ -0,0 +1,84 @@
+# Create training and test data
+np.random.seed(42)
+n_datapoints = 100
+C=[[0.9, 0.0],[0.0, 20.0]]
```
pep8

agramfort commented on examples/preprocessing/plot_robust_scaling.py:
```diff
@@ -0,0 +1,84 @@
+mu2 = [101.0, -3.0]
+X1 = np.random.multivariate_normal(mean=mu1, cov=C, size=n_datapoints)
+X2 = np.random.multivariate_normal(mean=mu2, cov=C, size=n_datapoints)
+Y_train = np.hstack([[-1]*n_datapoints, [1]*n_datapoints])
```
please run pep8 checker on this script

agramfort commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +399,239 @@ def inverse_transform(self, X, copy=None):
+class RobustScaler(BaseEstimator, TransformerMixin):
+    """Scale features using statistics that are robust to outliers.
+
+    This Scaler removes the median and and scales the data according to
```
and and

please go over the text one more time. I'll read again when you did so.

untom committed:
Fix documentation errors

untom commented:
@agramfort Thanks for taking the time to look at this. I've fixed the mistakes you pointed out, and went over the docstrings in `data.py` again and removed the repeated words/phrases. If you could look it over once more now, that'd be great.

agramfort commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +399,239 @@ def inverse_transform(self, X, copy=None):
+    Attributes
+    ----------
+    `center_` : array of floats
+        The median value for each feature in the training set, unless axis=1,
```
I don't see any axis parameter

untom replied:
You are right, that is a left-over from a feature that I later removed because an other reviewer didn't like it. Thanks for catching it!

agramfort commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +399,239 @@ def inverse_transform(self, X, copy=None):
+                scale = 1.
+        elif isinstance(scale, np.ndarray):
+            scale[scale == 0.0] = 1.0
+            scale[-np.isfinite(scale)] = 1.0
```
do you mean scale[~np.isfinite(scale)] with a ~?

GaelVaroquaux replied:
> do you mean scale[~np.isfinite(scale)] with a ~?

One extra reason for using 'np.logical_not'.

agramfort replied:
+1

did you check coverage? it's a sign tests are not complete...

untom replied:
There is a test to catch that. But it appears that on a binary array, - does the same thing as ~ does, so the test didn't catch this.

    y = np.random.normal(size=(4, 5)) > 0   # random binary array
    (-y == ~y).all()   # gives True

Anyway, it's fixed now.

agramfort commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +399,239 @@ def inverse_transform(self, X, copy=None):
+                if X.shape[0] == 1:
+                    inplace_row_scale(X, 1.0 / self.scale_)
+                elif self.axis == 0:
+                    inplace_column_scale(X, 1.0 / self.scale_)
```
so you can transform sparse but not fit with sparse?

untom replied:
Yes. The problem is that calculating the median (or IQR) of a sparse-matrix is quite non-trivial. So I didn't implement it for now. But I figured if someone wants to transform a sparse matrix after fitting a dense one, why not let them?

agramfort commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +399,239 @@ def inverse_transform(self, X, copy=None):
+        if self.with_scaling:
+            q = np.percentile(X, (25, 75), axis=0)
+            self.scale_ = (q[1] - q[0])
+            if np.isscalar(self.scale_):
```
how could this happen with X.shape == (n_samples, n_features) ?

untom replied:
This can happen if the input isn't actually "array-like with shape [n_samples, n_features]", but rather is a 1D array. `StandardScaler` behaves the same way, and uses the same docstring to describe X. The right thing to do would probably be to change the documentation of both StandardScaler and RobustScaler to somethign like "array-like with shape [n_samples, n_features] or 1D array". Is that a good formulation for it?

agramfort commented on sklearn/preprocessing/data.py:
```diff
@@ -397,6 +399,239 @@ def inverse_transform(self, X, copy=None):
+            if self.with_scaling:
+                X *= self.scale_
+            if self.with_centering:
+                X += self.center_
```
this is a big code duplication. Please factorize with private function or method

untom replied:
I tried, but couldn't come up with a good factorization. I noticed 4 redundant lines (`self.copy` is checked once within `check_array` and once within the method itself), but other than that, refactoring makes the code just more complex/adds a level of indirection, but barely saves any code. There are two redundant parts:

1. I could factor out the `check_is_fitted` calls and the `_check_array`. However, that results in a total of 1 saved line, which I don't think warants including a new method.
2. Scaling the sparse matrix is the same in both methods, so that could be factorized out. But again, that would save just 1 line.

I can't combine these two redundancies into a single method, because that would mean splitting up the `if sparse.issparse` check over two methods, which makes the code necessarily hard to understand.

untom committed:
More documentation fixes

untom committed:
Remove redundant code

agramfort commented:
ok LGTM

one last review?

untom commented:
Thanks for looking it over. Do PRs need 3 reviews before being accepted now?

agramfort commented:
no but I'd like to have one more review to see if my last comments would be
nitpicks for another core dev

GaelVaroquaux commented:
Ok, I find that the issues raised by @agramfort are not major ones. I think that this can be merged as such. 

However :), the median and the inter quartile range are very robust estimators but also very noisy. Using a trimmed mean, both on the raw data to estimate center location,  and on the squared distances to the center,  to estimate spread, is in general a much better strategy with a continuous parameter to interpolate between mean and median. 

I would really love a PR adding such a behavior to the robust scaler. It worries me that anybody wanting to do robust scaling will right now have also very noisy scaling.

untom commented:
I never heard of the trimmed mean being more robust, could you name any references I could read up on?

amueller commented:
Merged. Thanks @untom.
Would you like to do the MaxAbsScaler?

untom commented:
Nice! Thanks to the reviewers for their comments.

@amueller: I am fairly busy until the 5th, so most likely I will only find time to work on this afterwards.

amueller commented:
I smell nips ;) Good luck!

mblondel commented:
> In such cases, the median and the interquartile range often give better results.

If there are papers to support this claim, it would be nice to add them to the references.

ogrisel commented:
Thanks @untom for this contribution. I added a what's new entry here: 549ecaefe299e46bf32890a7554d1a2360961061.

Please feel free to open a new PR to update it to use your real name instead of your github nickname. You might also want to update the authors list in the header of the `sklearn/preprocessing/data.py` file.
